### PR TITLE
Update docs in order to avoid calling deprecated methods `setTemplate()` and `setTemplates()` in `AbstractAdmin`

### PR DESCRIPTION
**Pull request by phansys**
## Subject
Update docs about templates.

PR #6387 deprecates methods `AbstractAdmin::setTemplate()` and `AbstractAdmin::setTemplates()`, but the documentation still recommends to use them.
These calls were updated in order to use the template registry services.

I am targeting this branch, because these changes respect BC.

Related to #6387.

## Changelog

```markdown
### Fixed
- Deprecated calls to `AbstractAdmin::setTemplate()` and `AbstractAdmin::setTemplates()` in docs.
```

### CODE REVIEW

**Comment by greg0ire** [comment]
Documentation is not that different from code: it does deserve separate commits for grammar fixes and for other things. Also, it does deserve descriptive commit messages that explain what you are doing, not just which area of the docs you are working on. Typically, this should appear in your commit message:

> deprecates methods AbstractAdmin::setTemplate() and AbstractAdmin::setTemplates(), but the documentation still recommends to use them.

**Comment by phansys** [comment]
The commit and the PR description were updated. Please, let me know if it accomplish your expectations.

File: docs/cookbook/recipe_row_templates.rst [hunk 1]
```diff
@@ -17,7 +17,9 @@ The recipe
 Configure your Admin service
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-The configuration takes place in the DIC by calling the ``setTemplates`` method.
+The configuration takes place in the DIC by calling the ``setTemplate()`` / ``setTemplates()``
+method in the service "%s.template_registry" service, where "%s" is the name of your
+admin service (by instance "sonata.admin.comment").
```
Comment (greg0ire) [inline docs/cookbook/recipe_row_templates.rst:22]: Let's avoid strings that may contain markdown markup (here, the underscore)
```suggestion
method in the service`%s.template_registry` service, where `%s` is the name of your
admin service (by instance `sonata.admin.comment`).
```

File: docs/cookbook/recipe_customizing_a_mosaic_list.rst [hunk 2]
```diff
@@ -33,26 +33,15 @@ First, configure the ``outer_list_rows_mosaic`` template key:
...
-       </service>
-
-The ``list_outer_rows_mosaic.html.twig`` is the name of one mosaic's tile. You should also extends the template and overwrite the default blocks availables.
+                 <argument type="collection">
+                     <argument key="outer_list_rows_mosaic">@SonataMedia/MediaAdmin/list_outer_rows_mosaic.html.twig</argument>
+                 </argument>
+             </call>
+        </service>
+
+The ``list_outer_rows_mosaic.html.twig`` is the name of one mosaic's tile. You should also extend the template and overwrite the default blocks available.
```
Comment (greg0ire) [inline docs/cookbook/recipe_customizing_a_mosaic_list.rst:44]: That should be a separate commit

File: docs/cookbook/recipe_row_templates.rst [hunk 3]
```diff
@@ -85,6 +76,6 @@ Once the templates are defined, create the template to render the row:
     {% endblock %}
 
 While this feature is nice to generate a rich list, you can break the layout and
-admin features such as batch and object actions. It is best to familiarise yourself
+admin features such as batch and object actions. It is best to familiarize yourself
```
Comment (greg0ire) [inline docs/cookbook/recipe_row_templates.rst:79]: That should be in the same commit as the other grammar/spelling change

File: docs/reference/templates.rst [hunk 4]
```diff
@@ -202,19 +191,17 @@ can specify the templates to use in the ``Admin`` service definition:
 
 Changes made using the ``setTemplate()`` and ``setTemplates()`` methods
 override the customizations made in the configuration file, so you can specify
-a global custom template and then override that customization on a specific
+a global custom template and then override that customization for a specific
```
Comment (greg0ire) [inline docs/reference/templates.rst:194]: Separate commit here too

File: docs/reference/templates.rst [hunk 5]
```diff
@@ -202,19 +191,17 @@ can specify the templates to use in the ``Admin`` service definition:
...
-the templates registered through the configuration above. Through the method
-``getTemplate($name)`` of that class, you can access the templates set for
-that ``Admin``. The ``TemplateRegistry`` is available through ``$this->getTemplateRegistry()``
-within the ``Admin``. Using the service container the template registries can
-be accessed outside an ``Admin``. Use the ``Admin`` code + ``.template_registry``
-as the service ID (i.e. "app.admin.post" uses the Template Registry
-"app.admin.post.template_registry").
+the templates registered through the configuration above. The ``TemplateRegistry``
+is available through ``$this->getTemplateRegistry()`` within the ``Admin``.
+Using the service container the template registries can be accessed outside an ``Admin``.
```
Comment (greg0ire) [inline docs/reference/templates.rst:202]: That sentence lacks punctuation IMO, also it might be better not to mention the service container other one people might end up using `$container->get` over DI.
```suggestion
Outside an ``Admin``, you can obtain the template registries with their service ID.
```

File: docs/reference/templates.rst [hunk 6]
```diff
@@ -202,19 +191,17 @@ can specify the templates to use in the ``Admin`` service definition:
...
+Use the ``Admin`` code + ``.template_registry`` as the service ID (i.e.
+"app.admin.post" uses the template registry "app.admin.post.template_registry").
```
Comment (greg0ire) [inline docs/reference/templates.rst:204]: I'd use code markup here, but it's funny, [that's the second time I have this discussion today](https://github.com/doctrine/dbal/pull/4384#discussion_r511178109), so don't hesitate to tell me if you feel strongly otherwise.

```suggestion
`app.admin.post` uses the template registry `app.admin.post.template_registry`).
```

File: docs/reference/templates.rst (hunk 4 above)
Comment (phansys) [reply docs/reference/templates.rst:194]: I agree about the others are grammar fixes, but this is a change about the meaning between the previous and the current behavior.
The previous setting was required **on** a specific admin definition, while the current one is required **for** a specific admin definition.

Comment (greg0ire) [reply docs/reference/templates.rst:194]: Oh yes, you are definitely right, I overlooked that one 👍

File: docs/reference/templates.rst (hunk 6 above)
Comment (phansys) [reply docs/reference/templates.rst:204]: I think there is no established rule for the case where we need to refer to service names, parameter names, package names, etc. (since they are **names**, sometimes it feels natural to use double quotes and treat them a literal strings).
Personally, I try to be consistent with this, but in some cases it depends on the context, I think.
In the same reasoning, I try to use backticks when referencing code snippets (with partial or complete fragments, IE `AdminInterface::id()`).

Comment (greg0ire) [reply docs/reference/templates.rst:204]: FYI the discussion I linked resolved in using backticks (there was no debate, we needed both backticks _and_ quotes). I tend to use backticks for everything. So to me you can write "admin class", you can also write `Admin` class, but you can't write Admin class, because now you are using a technical name as it appears in the code, and names that appear in code may contain special characters that have meaning in markdown like underscores or stars, but also in normal language, for example dots. When you use backticks, a monospace font will be used, and the reader knows you are talking about something that will appear in the code, and not approximately referring to something using a high level term like "unit of work" or "repository interface".

**Review by VincentLanglet — APPROVED** [review]

**Comment by wbloszyk** [comment]
I busy today. I have to finish some of my commercial works. I will try to find some time for it tomorrow. IMO it is problem with template registry which is created by compiler pass and do not exist when config is loaded.

**Comment by phansys** [comment]
I agree with that suspicion.
Thank you.

**Comment by wbloszyk** [comment]
My idea is create one, global admin template registry and force people to create own (by override default) if they want override templates. 
```xml
<service id="sonata.media.admin.media.template_registry">
    <tag name="sonata.teplate_registry"/>
    <call method="setTemplates">
       ...
    </call>
</service>
<service id="sonata.media.admin.media">
    <tag name="sonata.admiun"/>
    <call method="setTemplateRegistry">
       <argument type="service" id="sonata.media.admin.media.template_registry"> 
    </call>
</service>
```

Im feature will be nice to add multiple registry for one admin. Support for role, destiny (chooseable show as table or thumbnail etc.)

**Comment by phansys** [comment]
I understand. But I also think that we shouldn't deprecate `AbstractAdmin::setTemplate()` in favor of this approach, if isn't working yet.

**Comment by jordisala1991** [comment]
> I understand. But I also think that we shouldn't deprecate `AbstractAdmin::setTemplate()` in favor of this approach, if isn't working yet.

Agree with that, IMO we should undeprecate this until a fully working solution can be provided, and only if it is better than simply calling setTemplate

**Comment by wbloszyk** [comment]
I will try do it tomorrow.

**Comment by wbloszyk** [comment]
Miss click :laughing:

**Comment by wbloszyk** [comment]
@phansys 
I found resolve for it. 
At first global template registry should be non changeable:
https://github.com/sonata-project/SonataAdminBundle/pull/6556

At second:
`MutableTemplateRegistry` should have `parent` property. (for global settings)

At third:
User will be able to create `MutableTemplateRegistry` by own. Compiler pass will create or call setters then.

**Comment by phansys** [comment]
>At third:
User will be able to create MutableTemplateRegistry by own. Compiler pass will create or call setters then.

Could you please share an example about the required DI configuration to customize an admin template?

IE:
>BEFORE:
>```yaml
>admin.foo:
>    class: App\FooAdmin
>    arguments: [null, App\Entity\Foo, null]
>    calls:
>        - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]    
>```
>AFTER:
>```yaml
># ...
>```

**Comment by wbloszyk** [comment]
Compiler pass will auto inject `admin_code.template_registry` to `admin_code` when `setTemplateRegistry` is not call. So user will be able to set it with custom name too. Set template will be able by arguments too. I want add parent template registry (to reference for global values) or add template manager.
 
 BEFORE:
 ```yaml
 admin.foo:
    class: App\FooAdmin
    arguments: [null, App\Entity\Foo, null]
    calls:
        - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]    
```
 
 AFTER:
 ```yaml
admin.foo.template_registry:
    class: Sonata\AdminBundle\Templating\MutableTemplatingRegistry # IMO this can be set by tag, then class will be optional
    tags: 
        - [name: "sonata.template_registry"]
    calls:
        - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]    
 ```

**Comment by phansys** [comment]
Currently, the custom definition of `admin.foo.template_registry` is being overridden by the default definition. What do you suggest to fix that?
I insist that, if we aren't able to provide a working solution with this approach, we MUST undeprecate the previous one https://github.com/sonata-project/SonataAdminBundle/pull/6538#issuecomment-719783784, https://github.com/sonata-project/SonataAdminBundle/pull/6538#issuecomment-719787551.

**Comment by jordisala1991** [comment]
IMO the steps shouldnt be: lets try to make it work and if we cant, undeprecate it. It should be the other way around. Lets undeprecate for now and try to provide a real alternative way of doing things, then deprecate again.

wdyt @sonata-project/contributors ?

**Comment by phansys** [comment]
I agree, that's what I'm asking for.

**Comment by wbloszyk** [comment]
Templates are set in 3 steps: global, admin config(like pager) and from call setTemplate(s). 

> IMO the steps shouldnt be: lets try to make it work and if we cant, undeprecate it. It should be the other way around. Lets undeprecate for now and try to provide a real alternative way of doing things, then deprecate again.
> 
> wdyt @sonata-project/contributors ?

It is possible by check `setTemplateRegistry` is call in compiler pass (like with `setTemplate` and `setTempaltes`). If yes then we should take templates from argument end reconfigure it.

Todo it in easiest way and add some improvement:
- I want change global template registry to non mutable version. 
- Add second `parent` argument to `TemplateRegistry` (and call it in `getTemplate(s)`)
- Set it by `sonata.admin.global_template_registry` when null 

Global templates will not be set, they will be taken from parent instead. This change will allow create some admin template structure. For example for every product with possibility to simply override it: 

```yaml
admin.product.template_registry:
   class: Sonata\AdminBundle\Templating\MutableTemplatingRegistry # IMO this can be set by tag, then class will be optional
   arguments: [{'edit' : 'admin/product/edit.html.twig'}, admin.sonata.global_template_registry ] # second argument can be set by compiler pass is null
    tags: 
       - [name: "sonata.template_registry"]
   calls:
       - [setTemplate, ['edit', 'admin/product/edit.html.twig']]   

admin.product.travel.template_registry:
   class: Sonata\AdminBundle\Templating\MutableTemplatingRegistry # IMO this can be set by tag, then class will be optional
   arguments: [{'show', 'admin/product/travel/show.html.twig'}, admin.product.template_registry ] # second argument can be set by compiler pass is null
   tags: 
       - [name: "sonata.template_registry"]
```

**Comment by phansys** [comment]
It seems that we are discussing 2 different thing here.
IMO, we must focus on fix the current situation: we've deprecated an existing feature indicating that other approach must be followed, and that approach isn't working. So, we must undeprecate the previous working feature.
When that issue is solved, we can invest our effort in providing the new feature in a proper way (adding the required tests to ensure it is working, this time).

Could you please confirm that's how you've set your priorities about this?

**Comment by wbloszyk** [comment]
@phansys IMO this should resolve problem with config. Can you check it? https://github.com/sonata-project/SonataAdminBundle/pull/6561

**Comment by VincentLanglet** [comment]
> It seems that we are discussing 2 different things here.
> IMO, we must focus on fix the current situation: we've deprecated an existing feature indicating that other approach must be followed, and that approach isn't working. So, we must undeprecate the previous working feature.
> When that issue is solved, we can invest our effort in providing the new feature in a proper way (adding the required tests to ensure it is working, this time).
> 
> Could you please confirm that's how you've set your priorities about this?

After reading all this discussion, I just feel like this feature is not ready and was maybe not discuss enough.
IMHO this should not be in the next major (which we try to release ASAP).

This PR:
- https://github.com/sonata-project/SonataAdminBundle/pull/6387
seems to be the first of many
- https://github.com/sonata-project/SonataAdminBundle/pull/6556
- https://github.com/sonata-project/SonataAdminBundle/pull/6393
- https://github.com/sonata-project/SonataAdminBundle/pull/6467 

IMHO, we should just revert it and delay these changes for 5.0 if it's not ready and fully working.

Plus now I (maybe) start to understand it, I don't see the interest in this feature.
The change
```
admin.foo:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]  
```
to
```
admin.foo:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplateRegistry, '@admin.foo.template_registry']

admin.foo.template_registry:
   class: Sonata\AdminBundle\Templating\MutableTemplatingRegistry # IMO this can be set by tag, then class will be optional
   tags: 
       - [name: "sonata.template_registry"]
   calls:
       - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]
```
does not seem a good thing to me.

Setting a templating is actually so easy that it doesn't need a service for this.
```
admin.foo:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplateRegistry, '@admin.foo.template_registry']

admin.foo_bar:
   class: App\FooBarAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplateRegistry, '@admin.foo.template_registry']

admin.foo.template_registry:
   class: Sonata\AdminBundle\Templating\MutableTemplatingRegistry # IMO this can be set by tag, then class will be optional
   tags: 
       - [name: "sonata.template_registry"]
   calls:
       - [setTemplate, ['show', 'admin/foo/show.html.twig']]
       - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]
```
Can just be done this way
```
admin.foo:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplate, ['show', 'admin/foo/show.html.twig']]  
       - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]

admin.foo_bar:
   class: App\FooBarAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplate, ['show', 'admin/foo/show.html.twig']]  
       - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]
```
and I find this simpler.

Personally when I override show/edit template, it's alway for one single Admin. Asking me to create a service everytime I need to override a template would be counter productive. Did I missunderstood something @wbloszyk @phansys ?

Maybe both way should co-exist:
- if you want to override template for one admin: use setTemplate
- if you want to override template for multiple admin: use service

**Comment by phansys** [comment]
>@phansys IMO this should resolve problem with config. Can you check it? #6561

I can confirm it isn't working. You could check locally by introspecting the compiled debug container in your environment (`var/cache/{env}/Kernel{env}DebugContainer.xml`) and searching for `admin.product.template_registry`.

>Personally when I override show/edit template, it's alway for one single Admin. Asking me to create a service everytime I need to override a template would be counter productive. Did I missunderstood something @wbloszyk @phansys ?

I understand exactly the same. Please, note that I've created this PR by guessing what was done at #6387, but I'm not have certainty about the required steps to replace the deprecated feature.
I agree that `AbstractAdmin::setTemplate()` seems much more simpler than the approach introduced in that PR.

**Comment by wbloszyk** [comment]
#6556 should in 3.x. Other step in master or in 4.x.

IMO keep this methods is bad decision. Templates should be manage by templates registry. But we can keep calling setters by catching this args, inject it by argument and remove and remove calling in compiler pass. 

WDYT?

**Comment by VincentLanglet** [comment]
> IMO keep this methods is bad decision. Templates should be manage by templates registry. 

Why should it be ?
Creating a service to change the templates of multiple admins is not a bad idea.
But creating a service for changing one template seems over-engineered. 

> But we can keep calling setters by catching this args, inject it by argument and remove and remove calling in compiler pass.

I'm not sure to fully understand. Do you want to auto-generate a templateRegistry if we provide `setTemplate` ?

Can you provide a way to have both this config
```
admin.foo:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplateRegistry, '@admin.foo.template_registry']

admin.foo.template_registry:
   class: Sonata\AdminBundle\Templating\MutableTemplatingRegistry # IMO this can be set by tag, then class will be optional
   tags: 
       - [name: "sonata.template_registry"]
   calls:
       - [setTemplate, ['show', 'admin/foo/show.html.twig']]
       - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]
```
and this config 
```
admin.foo:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplate, ['show', 'admin/foo/show.html.twig']]  
       - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]
```
working ?

**Comment by VincentLanglet** [comment]
I just thought about a special case 

```
admin.foo1:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplate, ['show', 'admin/foo/show.html.twig']]  
       - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]

admin.foo2:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]

admin.foo3:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplate, ['show', 'admin/foo/show.html.twig']]
```

With template registry, you need to create 3 services because admin.foo1 can't set multiple templateRegistry.
Wouldn't it be better to use a collection of templateRegistry ?

```
admin.foo1:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [addTemplateRegistry, '@admin.foo.template_registry2']
       - [addTemplateRegistry, '@admin.foo.template_registry3']

admin.foo2:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [addTemplateRegistry, '@admin.foo.template_registry2']

admin.foo3:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [addTemplateRegistry, '@admin.foo.template_registry3']
```

**Comment by phansys** [comment]
None of the proposed solutions (#6556, #6561) fix the issue, so until it is resolved, I must rollback the latest release I made in one of my projects, and declare a conflict against `sonata-project/admin-bundle:>=3.76.0`.

**Comment by wbloszyk** [comment]
> > IMO keep this methods is bad decision. Templates should be manage by templates registry.
> 
> Why should it be ?
> Creating a service to change the templates of multiple admins is not a bad idea.

IMO keep `AbstractAdmin::setTemlate(s)` is bad idea. Templates should be manage by template registry. This will keep model more clear. 

> But creating a service for changing one template seems over-engineered.

Agree. That why we should keep `setTemplate(s)` only in config. Compiler pass will delete calls for them and move call templates to `TemplateRegistry` argument.  (hack)

> > But we can keep calling setters by catching this args, inject it by argument and remove and remove calling in compiler pass.
> 
> I'm not sure to fully understand. Do you want to auto-generate a templateRegistry if we provide `setTemplate` ?

`TemplateRegistry` can be create by user or will be auto-generated. In both way templates from `setTemplate(s)` will override `TemplateRegistry` argument.

> Can you provide a way to have both this config

I can provide something like this:

```
admin.foo:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplateRegistry, '@admin.foo.template_registry'] # optional
       - [setTemplate, ['edit', 'admin/bar/edit.html.twig']] # this template will override templates from template registry

admin.foo.template_registry:
   class: Sonata\AdminBundle\Templating\MutableTemplatingRegistry # IMO this can be set by tag, then class will be optional
   tags: 
       - [name: "sonata.template_registry"]
   calls:
       - [setTemplate, ['show', 'admin/foo/show.html.twig']]
       - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]
```

**Comment by wbloszyk** [comment]
> None of the proposed solutions (#6556, #6561) fix the issue, so until it is resolved, I must rollback the latest release I made in one of my projects, and declare a conflict against `sonata-project/admin-bundle:>=3.76.0`.

@phansys I can fix in the evening if you agree with https://github.com/sonata-project/SonataAdminBundle/pull/6538#issuecomment-720436410

**Comment by phansys** [comment]
>@phansys I can fix in the evening if you agree with #6538 (comment)

I agree that it could be a working solution, but I'm in doubt since the template registry approach was not working in any of the related PRs yet.
Regarding the hack of removing the call to `setTemplate()` and `setTemplates()` in the admin service definition and passing them to the template registry service, I think we should chose a different name for these methods in the service calls to avoid confusions and make the situation a little bit clear. Maybe something like `registerTemplate()` or something that isn't a real method name in the admin class.

**Comment by VincentLanglet** [comment]
> > But creating a service for changing one template seems over-engineered.
> 
> Agree. That why we should keep `setTemplate(s)` only in config. Compiler pass will delete calls for them and move call templates to `TemplateRegistry` argument. (hack)

Not sure if this hack will be natural for developer who won't understand why adding a `setTemplate` is allowed in `calls` when there is no `setTemplate` method in the AbstractAdmin class.

> Regarding the hack of removing the call to setTemplate() and setTemplates() in the admin service definition and passing them to the template registry service, I think we should chose a different name for these methods in the service calls to avoid confusions and make the situation a little bit clear. Maybe something like registerTemplate() or something that isn't a real method name in the admin class.

I agree. Best would be a `templates` key in the config, but I assume it's not possible ?

> > Can you provide a way to have both this config
> 
> I can provide something like this:
> 
> ```
> admin.foo:
>    class: App\FooAdmin
>    arguments: [null, App\Entity\Foo, null]
>    calls:
>        - [setTemplateRegistry, '@admin.foo.template_registry'] # optional
>        - [setTemplate, ['edit', 'admin/bar/edit.html.twig']] # this template will override templates from template registry
> 
> admin.foo.template_registry:
>    class: Sonata\AdminBundle\Templating\MutableTemplatingRegistry # IMO this can be set by tag, then class will be optional
>    tags: 
>        - [name: "sonata.template_registry"]
>    calls:
>        - [setTemplate, ['show', 'admin/foo/show.html.twig']]
>        - [setTemplate, ['edit', 'admin/foo/edit.html.twig']]
> ```

Seems great

> Update for @VincentLanglet
> I want add `addTemplateRegistry` to allow use multiple registry in 5.0 milestone.

Ok.

**Comment by phansys** [comment]
>I agree. Best would be a templates key in the config, but I assume it's not possible ?

Maybe we could transmit the "magic" concept by using `__templates` as method name.

**Comment by wbloszyk** [comment]
> > I agree. Best would be a templates key in the config, but I assume it's not possible ?
> 
> Maybe we could transmit the "magic" concept by using `__templates` as method name.

:thinking: 

I found the best solution:
https://symfony.com/doc/current/service_container/tags.html#adding-additional-attributes-on-tags

We can use additional `templates` atributes on `sonata.template_registry` tag.

**Comment by VincentLanglet** [comment]
> We can use additional `templates` atributes on `sonata.template_registry` tag.

Can you give an example of the config you have in mind @wbloszyk ?

If it's on `sonata.template_registry` I'm not sure it resolve our concern about the `AbstractAdmin::setTemplate` method
```
admin.foo:
   class: App\FooAdmin
   arguments: [null, App\Entity\Foo, null]
   calls:
       - [setTemplate, ['edit', 'admin/bar/edit.html.twig']]
```

**Comment by wbloszyk** [comment]
> > We can use additional `templates` atributes on `sonata.template_registry` tag.
> 
> Can you give an example of the config you have in mind @wbloszyk ?
> 
> If it's on `sonata.template_registry` I'm not sure it resolve our concern about the `AbstractAdmin::setTemplate` method
> 
```
admin.foo:
    class: App\FooAdmin
        tags: 
            - [name: "sonata.template_registry", templates: ['edit': 'admin/bar/edit.html.twig', 'show', 'admin/bar/show.html.twig'] ]
        arguments: [null, App\Entity\Foo, null]
 ```

Of course in 3.x `AbstractAdmin::templates` property are used. In this case we have to set it by `AbstractAdmin`.

**Comment by VincentLanglet** [comment]
> ```
> admin.foo:
>     class: App\FooAdmin
>         tags: 
>             - [name: "sonata.template_registry", templates: ['edit': 'admin/bar/edit.html.twig', 'show': 'admin/bar/show.html.twig'] ]
>         arguments: [null, App\Entity\Foo, null]
> ```

Seems great.
Could you implement this @wbloszyk ?

With this, I think we don't have to revert the deprecation @phansys.

**Comment by phansys** [comment]
>With this, I think we don't have to revert the deprecation @phansys.

Agreed.

**Comment by wbloszyk** [comment]
> Seems great.
> Could you implement this @wbloszyk ?

I will try to do it today.

Additional this will add the same functionality to `MutableTemplateRegistryAwereInterface`.

**Comment by wbloszyk** [comment]
Some update: https://github.com/sonata-project/SonataAdminBundle/pull/6566

- tag attribute can not be array, I used `virtual_template_method` instead
- I have some problem with current `AddDependencyCallsCompilerPass:fixTemplates()` but I found solution


Thursday is family day. I will finish it in Friday.

**Comment by kirya-dev** [comment]
Guys. Whyt `AbstractAdmin::getTemplates()` that

> > ```
> > admin.foo:
> >     class: App\FooAdmin
> >         tags: 
> >             - [name: "sonata.template_registry", templates: ['edit': 'admin/bar/edit.html.twig', 'show': 'admin/bar/show.html.twig'] ]
> >         arguments: [null, App\Entity\Foo, null]
> > ```
> 
> 
> Seems great.
> 
> Could you implement this @wbloszyk ?

Me seams that its too complex to programming.
Configuring sonata admin job for very hard programmers?

**Comment by kirya-dev** [comment]
Whyt abount configure templates in php code?

In `FooBarAdmin::configure`  make something:
```php
    public function configure(): void
    {
        $this->setTemplates([
            'list' => '@FooBar/list.html.twig',
        ]);
    }
```

**Comment by kirya-dev** [comment]
Or make it another:
```
    public function getTemplates(): array
    {
        return [
            'list' => '@FooBar/list.html.twig',
        ];
    }
```

**Comment by kirya-dev** [comment]
Also. in this PR impossible configure templates with fully autowiring admins.

**Comment by VincentLanglet** [comment]
> What about configure templates in php code?
> 
> In `FooBarAdmin::configure` make something: (see symfony 5 way - zero yml configuration for new services)
> 
> ```
> <?php
> declare(strict_types=1);
> 
> namespace App\Admin;
> 
> use App\Entity\FooBar;
> 
> class FooBarAdmin extends AbstractAdmin
> {
>     public function __construct()
>     {
>         parent::__construct('foo-bar', FooBar::class);
>     }
> 
>     public function configure(): void
>     {
>         $this->getTemplateRegistry()->setTemplates([
>             'list' => '@FooBar/list.html.twig',
>         ]);
>     }
> }
> ```

You can do it with configure indeed. But it's not bad to allow user to use `yaml` too.
Everybody will use his prefer way.

**Comment by kirya-dev** [comment]
> > What about configure templates in php code?
> > In `FooBarAdmin::configure` make something: (see symfony 5 way - zero yml configuration for new services)
> > ```
> > <?php
> > declare(strict_types=1);
> > 
> > namespace App\Admin;
> > 
> > use App\Entity\FooBar;
> > 
> > class FooBarAdmin extends AbstractAdmin
> > {
> >     public function __construct()
> >     {
> >         parent::__construct('foo-bar', FooBar::class);
> >     }
> > 
> >     public function configure(): void
> >     {
> >         $this->getTemplateRegistry()->setTemplates([
> >             'list' => '@FooBar/list.html.twig',
> >         ]);
> >     }
> > }
> > ```
> 
> You can do it with configure indeed. But it's not bad to allow user to use `yaml` too.
> Everybody will use his prefer way.

Its possible add my example to documentation?

**Comment by kirya-dev** [comment]
Whyt ?
```php
    protected function configureTemplateRegistry(TemplateRegistry $registry): void
    {
        $registry->setTemplate('edit', '@FooBar/FooBarAdmin/edit.html.twig');
    }
```

**Comment by kirya-dev** [comment]
Also found problem: When call `setTemplates()` this fully discard previous (global) templates

**Comment by VincentLanglet** [comment]
There is no real need to discuss about this PR before https://github.com/sonata-project/SonataAdminBundle/pull/6566

**Comment by phansys** [comment]
@wbloszyk, could I close this PR in favor of #6566? I think you can add the required documentation there.
